vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/dao/impl/DeviceRepositoryImpl.java

```diff
@@ -0,0 +1,175 @@
+            objectInputStream = new ObjectInputStream(new FileInputStream(file));
+            Map<String, DeviceInfo> deviceInfoMap = (Map<String, DeviceInfo>) objectInputStream.readObject();
+            reLogin(deviceInfoMap);
+            CACHE_DEVICE_INFO.putAll(deviceInfoMap);
+        } catch (Exception e) {
+            logger.warn(e.getMessage(),e);
+        } finally {
+            quietClose(objectInputStream);
+        }
+    }
+
+    @Override
+    public Map<String, DeviceInfo> getAll() {
+        return CACHE_DEVICE_INFO;
+    }
+
+    @Override
+    public void remove(String uid) {
+        DeviceInfo deviceInfo = CACHE_DEVICE_INFO.remove(uid);
+        LOGINED_DEVICE.remove(deviceInfo.ip);
+        writeDataToFile();
+    }
+
+    @Override
+    public void update(String uid, DeviceInfo deviceInfo) throws Exception {
+        initDevice(deviceInfo);
+        CACHE_DEVICE_INFO.put(uid, deviceInfo);
+        writeDataToFile();
+    }
+
+    @Override
+    public void add(String uid, DeviceInfo deviceInfo) throws Exception {
+        initDevice(deviceInfo);
+        CACHE_DEVICE_INFO.put(uid, deviceInfo);
+        writeDataToFile();
+    }
+
+
+    @Override
+    public DeviceInfo get(String uid) {
+        return CACHE_DEVICE_INFO.get(uid);
+    }
+
+    @Override
+    public Storage getLoginedDeviceByIP(String deviceIP) {
+        return LOGINED_DEVICE.get(deviceIP);
+    }
+
+
+    private void writeDataToFile() {
+        ObjectOutputStream objectOutputStream = null;
+        try {
+            File file = new File(DEVICE_DATA_FILE_PATH);
+            objectOutputStream = new ObjectOutputStream(new FileOutputStream(file));
+            synchronized (CACHE_DEVICE_INFO) {
+                objectOutputStream.writeObject(CACHE_DEVICE_INFO);
+            }
+        } catch (Exception e) {
+            logger.error(e.getMessage(),e);
+        } finally {
+            quietClose(objectOutputStream);
+        }
+    }
+
+    private void initDevice(DeviceInfo deviceInfo) throws Exception {
+        Storage device = StorageFactory.newStorage(getDeviceClassName(deviceInfo.deviceType), "");
+        device.login(deviceInfo.ip, deviceInfo.port, deviceInfo.username, deviceInfo.password);
+        StorageMO deviceMO = device.getDeviceInfo();
+        deviceInfo.deviceModel = deviceMO.model;
+        deviceInfo.deviceName = deviceMO.name;
+        deviceInfo.deviceStatus = deviceMO.status;
+        deviceInfo.sn = deviceMO.sn;
+        LOGINED_DEVICE.put(deviceInfo.ip, device);
+    }
+
+    private String getDeviceClassName(String deviceType) {
+        String[] types = StorageFactory.listStorages();
+        for (String type : types) {
+            if (type.substring(type.lastIndexOf(".") + 1).equals(deviceType)) {
+                return type;
+            }
+        }
+        return "";
+    }
+
+    private void quietClose(Closeable closeable) {
+        if (closeable != null) {
+            try {
+                closeable.close();
```

Comment:
There won't be any concurrency issue?

Reply:
@kumarashit Could you please elaborate for concurrency issue